File: lambdas/nodejs/light/handler.ts
import { BatchGetItemCommand, BatchWriteItemCommand, DynamoDBClient } from "@aws-sdk/client-dynamodb";
import type { Context } from "aws-lambda";

const dynamoClient = new DynamoDBClient({});

interface BenchmarkSuccess {
  success: true;
  workloadType: 'light';
  architecture: string;
  memoryLimitMB: number;
  itemsWritten: number;
  itemsRead: number;
  writeRequestId: string;
  readRequestId: string;
  allDataMatches: boolean;
}

interface BenchmarkError {
  success: false;
  workloadType: 'light';
  error: string;
}

type BenchmarkResult = BenchmarkSuccess | BenchmarkError;

/**
 * Lambda handler - Light test performs a DynamoDB batch write (5 items) followed by a batch read to measure
 * baseline Lambda invocation and SDK initialization overhead with realistic
 * multi-item I/O patterns. Returns simple object for direct Lambda invocation.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export async function handler(_event: any, context: Context): Promise<BenchmarkResult> {
  console.log(
    JSON.stringify({
      event: 'handler_start',
      workloadType: 'light',
      runtime: `nodejs${process.version}`,
      architecture: process.arch,
      requestId: context.awsRequestId
    })
  );

  try {
    const result = await lightWorkload();

    console.log(
      JSON.stringify({
        event: 'handler_success',
        itemsWritten: result.itemsWritten,
        itemsRead: result.itemsRead,
        writeRequestId: result.writeRequestId,
        readRequestId: result.readRequestId,
        allDataMatches: result.allDataMatches
      })
    );

    return {
      success: true,
      workloadType: 'light',
      architecture: process.arch,
      memoryLimitMB: parseInt(context.memoryLimitInMB, 10),
      itemsWritten: result.itemsWritten,
      itemsRead: result.itemsRead,
      writeRequestId: result.writeRequestId,
      readRequestId: result.readRequestId,
      allDataMatches: result.allDataMatches
    };
  } catch (error) {
    console.error(
      JSON.stringify({
        event: 'handler_error',
        errorType: error instanceof Error ? error.constructor.name : 'Unknown',
        errorMessage: error instanceof Error ? error.message : String(error)
      })
    );

    return {
      success: false,
      workloadType: 'light',
      error: error instanceof Error ? error.message : String(error)
    };
  }
}

async function lightWorkload(): Promise<{
  writeRequestId: string;
  readRequestId: string;
  itemsWritten: number;
  itemsRead: number;
  allDataMatches: boolean;
}> {
  const table = process.env.DYNAMODB_TABLE_NAME || 'benchmark-test-data';

  const timestamp = Date.now();
  const arch = process.arch;
  const nodeVersion = process.version;
  const ttlTimestamp = Math.floor(Date.now() / 1000) + 86400; // 24 hours from now

  const items = [];
  for (let i = 0; i < 5; i++) {
    const itemId = `test-${timestamp}-${i}`;
    const data = `benchmark test data - nodejs${nodeVersion} ${arch} - item ${i}`;
    items.push({
      itemId,
      data,
      item: {
        pk: { S: itemId },
        sk: { S: 'light' },
        timestamp: { N: (timestamp + i).toString() },
        ttl: { N: ttlTimestamp.toString() },
        workload: { S: 'light' },
        runtime: { S: `nodejs${nodeVersion}` },
        architecture: { S: arch },
        data: { S: data }
      }
    });
  }

  const batchWriteCommand = new BatchWriteItemCommand({
    RequestItems: {
      [table]: items.map(item => ({
        PutRequest: {
          Item: item.item
        }
      }))
    }
  });

  const writeResponse = await dynamoClient.send(batchWriteCommand);

  const batchGetCommand = new BatchGetItemCommand({
    RequestItems: {
      [table]: {
        Keys: items.map(item => ({
          pk: { S: item.itemId },
          sk: { S: 'light' }
        }))
      }
    }
  });

  const readResponse = await dynamoClient.send(batchGetCommand);

  if (!readResponse.Responses || !readResponse.Responses[table]) {
    throw new Error(`No responses from table: ${table}`);
  }

  const retrievedItems = readResponse.Responses[table];
  if (retrievedItems.length !== items.length) {
    throw new Error(`Expected ${items.length} items, got ${retrievedItems.length}`);
  }

  const itemsById = new Map<string, string>();
  for (const retrievedItem of retrievedItems) {
    const itemId = retrievedItem.pk?.S || '';
    const data = retrievedItem.data?.S || '';
    itemsById.set(itemId, data);
  }

  const allDataMatches = items.every(item => {
    const retrievedData = itemsById.get(item.itemId);
    return retrievedData === item.data;
  });

  return {
    writeRequestId: writeResponse.$metadata.requestId || 'unknown',
    readRequestId: readResponse.$metadata.requestId || 'unknown',
    itemsWritten: items.length,
    itemsRead: retrievedItems.length,
    allDataMatches
  };
}

File: lambdas/python/light/handler.py
from __future__ import annotations

import json
import logging
import os
import platform
import time
from typing import Any

import boto3

logger = logging.getLogger()
logger.setLevel(logging.INFO)

DEFAULT_TABLE = "benchmark-test-data"
dynamodb = boto3.client("dynamodb")


def lambda_handler(_event: dict[str, Any] | None, context) -> dict[str, Any]:
    """Lambda handler - Light workload benchmark.

    Performs a DynamoDB batch write (5 items) followed by a batch read to measure
    baseline Lambda invocation and SDK initialization overhead with realistic
    multi-item I/O patterns.
    """
    logger.info(json.dumps({
        "event": "handler_start",
        "workloadType": "light",
        "runtime": f"python{platform.python_version()}",
        "architecture": platform.machine(),
        "requestId": getattr(context, "aws_request_id", "unknown")
    }))

    try:
        write_result = _write_batch()

        # Batch read back the same items to verify round-trip
        read_result = _read_batch(write_result["itemIds"])

        # Match items by ID (batch_get_item doesn't guarantee order)
        all_match = all(
            read_result["items"].get(item["itemId"]) == item["data"]
            for item in write_result["items"]
        )

        logger.info(json.dumps({
            "event": "handler_success",
            "itemsWritten": len(write_result["items"]),
            "itemsRead": len(read_result["items"]),
            "writeRequestId": write_result["requestId"],
            "readRequestId": read_result["requestId"],
            "allDataMatches": all_match
        }))

        return {
            "success": True,
            "workloadType": "light",
            "architecture": platform.machine(),
            "pythonVersion": platform.python_version(),
            "memoryLimitMB": int(getattr(context, "memory_limit_in_mb", 0) or 0),
            "itemsWritten": len(write_result["items"]),
            "itemsRead": len(read_result["items"]),
            "writeRequestId": write_result["requestId"],
            "readRequestId": read_result["requestId"],
            "allDataMatches": all_match,
        }
    except Exception as e:
        logger.error(json.dumps({
            "event": "handler_error",
            "errorType": type(e).__name__,
            "errorMessage": str(e)
        }))
        return _fail(f"{type(e).__name__}: {e}")


def _write_batch() -> dict[str, Any]:
    """Batch write 5 test items to DynamoDB with 24-hour TTL."""
    table = os.environ.get("DYNAMODB_TABLE_NAME") or DEFAULT_TABLE
    now_ms = int(time.time() * 1000)
    arch = platform.machine()  # aarch64 or x86_64
    python_ver = platform.python_version()
    ttl = int(time.time()) + 86_400  # 24h in seconds

    # Create 5 items with unique IDs
    items = []
    for i in range(5):
        item_id = f"test-{now_ms}-{i}"
        data = f"benchmark test data - python{python_ver} {arch} - item {i}"
        items.append({
            "itemId": item_id,
            "data": data,
            "item": {
                "pk": {"S": item_id},
                "sk": {"S": "light"},
                "timestamp": {"N": str(now_ms + i)},
                "ttl": {"N": str(ttl)},
                "workload": {"S": "light"},
                "runtime": {"S": f"python{python_ver}"},
                "architecture": {"S": arch},
                "data": {"S": data},
            }
        })

    # Batch write all items
    request_items = {
        table: [{"PutRequest": {"Item": item["item"]}} for item in items]
    }
    resp = dynamodb.batch_write_item(RequestItems=request_items)

    return {
        "requestId": resp["ResponseMetadata"]["RequestId"],
        "itemIds": [item["itemId"] for item in items],
        "items": [{"itemId": item["itemId"], "data": item["data"]} for item in items]
    }


def _read_batch(item_ids: list[str]) -> dict[str, Any]:
    """Batch read 5 test items from DynamoDB to verify write."""
    table = os.environ.get("DYNAMODB_TABLE_NAME") or DEFAULT_TABLE

    # Batch read all items
    keys = [{"pk": {"S": item_id}, "sk": {"S": "light"}} for item_id in item_ids]
    resp = dynamodb.batch_get_item(
        RequestItems={
            table: {"Keys": keys}
        }
    )

    if table not in resp.get("Responses", {}):
        msg = f"No responses from table: {table}"
        raise ValueError(msg)

    items = resp["Responses"][table]
    if len(items) != len(item_ids):
        msg = f"Expected {len(item_ids)} items, got {len(items)}"
        raise ValueError(msg)

    # Create a map of itemId -> data for matching (batch_get_item doesn't guarantee order)
    items_by_id = {
        item.get("pk", {}).get("S", ""): item.get("data", {}).get("S", "")
        for item in items
    }

    return {
        "requestId": resp["ResponseMetadata"]["RequestId"],
        "items": items_by_id
    }


def _fail(msg: str) -> dict[str, Any]:
    """Return error response in standard format."""
    return {"success": False, "workloadType": "light", "error": msg}
